allow adding a MDS to already deployed cluster

rishabh-d-dave opened the pull request:
Add a playbook that creates a new MDS node as a part of already
deployed Ceph cluster and deploys MDS there.

rishabh-d-dave commented:
TODO:
* make add_mds work for container scenario

rishabh-d-dave commented:
Attaching logs where `dev-centos-non_container-add_mdss` and `dev-centos-container-add_mdss` completed running successfully.

[dev-centos-non_container-add_mdss-SUCCEEDED.log](https://github.com/ceph/ceph-ansible/files/3008252/dev-centos-non_container-add_mdss-SUCCEEDED.log)
[dev-centos-container-add_mdss-SUCCEEDED.log](https://github.com/ceph/ceph-ansible/files/3008253/dev-centos-container-add_mdss-SUCCEEDED.log)
[dev-centos-non_container-add_mdss.log](https://github.com/ceph/ceph-ansible/files/3008254/dev-centos-non_container-add_mdss.log)
[dev-centos-container-add_mdss.log](https://github.com/ceph/ceph-ansible/files/3008255/dev-centos-container-add_mdss.log)

guits commented:
jenkins test dev-centos-non_container-add_mdss

guits commented on infrastructure-playbooks/add-mds.yml:
```diff
@@ -0,0 +1,67 @@
+    - import_role:
+        name: ceph-mds
+
+# update config files on OSD nodes
```
if this is not needed, please remove these lines.

rishabh-d-dave replied:
Done.

guits commented on infrastructure-playbooks/add-mds.yml:
```diff
@@ -0,0 +1,67 @@
+        name: ceph-infra
+    - import_role:
+        name: ceph-mds
+
```
we could probably print the output of `ceph status` to show the cluster state after the operation.

rishabh-d-dave replied:
Done.

guits commented on infrastructure-playbooks/add-mds.yml:
```diff
@@ -0,0 +1,67 @@
+      run_once: true
+      when: delegate_facts_host | bool
+  tasks:
+    - import_role:
```
you might want to import that role early in `pre_tasks` so you can use `mon_group_name` instead of hardcoding it at L21

rishabh-d-dave replied:
Done.

dsavineau commented:
out of my curiosity, does running the main playbook with `--limit mdss` would do the same than this standalone playbook ?

rishabh-d-dave commented on tox.ini:
```diff
@@ -169,6 +169,23 @@ commands=
+      ceph_dev_branch={env:CEPH_DEV_BRANCH:master} \
+      ceph_dev_sha1={env:CEPH_DEV_SHA1:latest} \
+      "
+  testinfra -n 4 --sudo -v --connection=ansible --ansible-inventory={changedir}/hosts-2 {toxinidir}/tests/functional/tests
```
This needs to replaced by `py.test` straightaway.

rishabh-d-dave replied:
Done.

rishabh-d-dave commented:
> out of my curiosity, does running the main playbook with --limit mdss would do the same than this standalone playbook ?

If the inventory file is not updated I think it would redploy MDS on `mdss` nodes but I can't be sure if the inventory files were updated it would have the same effect as running `add-mdss.yml`.

rishabh-d-dave commented:
Looks like it worked! I ran `dev-centos-non_container`with [this patch](https://github.com/ceph/ceph-ansible/files/3009370/before.patch.log) to setup the initial cluster (see [this attachment](https://github.com/ceph/ceph-ansible/files/3009359/dev-centos-non_container-add_mdss-before.log) for output) and then reversed the earlier patch and applied [this one](https://github.com/ceph/ceph-ansible/files/3009371/after.patch.log) and ran the same command again to add a mds node ([see this one](https://github.com/ceph/ceph-ansible/files/3009360/dev-centos-non_container-add_mdss-after.log) for output). Cluster looks fine to me -

```
[vagrant@mon0 ~]$ sudo ceph -k /var/lib/ceph/mon/ceph-mon0/keyring -n mon. status
  cluster:
    id:     efdb5a56-8018-41e3-a46f-d5e66638e57a
    health: HEALTH_WARN
            too few PGs per OSD (8 < min 30)
 
  services:
    mon: 1 daemons, quorum mon0 (age 28m)
    mgr: mon0(active, since 33m)
    mds: cephfs:1 {0=mds0=up:active}
    osd: 2 osds: 2 up (since 26m), 2 in (since 37m)
 
  data:
    pools:   2 pools, 16 pgs
    objects: 22 objects, 2.2 KiB
    usage:   27 GiB used, 35 GiB / 62 GiB avail
    pgs:     16 active+clean
 
[vagrant@mon0 ~]$ 
```

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
@guits @dsavineau 
There was only one error last time on CI. I don't think [it was](https://2.jenkins.ceph.com/job/ceph-ansible-prs-dev-centos-container-collocation/247/consoleFull#-155115346383b7c9d0-3feb-4c52-af68-464f9bda5975) due to my patch.

rishabh-d-dave commented:
@dsavineau @guits 
I've added new commit. What do you think about the changes in it? The intention is to not to run the playbook on the nodes where MDS is already deployed. If we agree that the new changes are better, I will go ahead and squash it.

dsavineau commented:
I don't think this is a great idea to create new groups with _old and _new prefix.
If you want to run this playbook only on new mds node the best way is probably to use the ansible `--limit` option

rishabh-d-dave commented:
That too would make the playbook run on nodes where MDS is already installed.

guits commented:
@rishabh-d-dave running ceph-ansible on a node which is already deployed shouldn't be a concern because it's idempotent, and if you concern is about time consumption, tasks are executed in parallel (unless you are deploying tons of MDS nodes...).

Why using `--limit` would make ceph-ansible running on mds nodes already deployed? You can pass node name to `--limit` instead of a group name.

rishabh-d-dave commented:
changes:
* removed the code that separates new and old mds nodes
* did a rebase and resolved a merge conflict

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-container-add_mdss

rishabh-d-dave commented:
[`dev-centos-non_container-add-mdss`](https://2.jenkins.ceph.com/job/ceph-ansible-prs-dev-centos-non_container-add_mdss/10/) job and  [`dev-centos-container-add_mdss`](https://2.jenkins.ceph.com/job/ceph-ansible-prs-dev-centos-container-add_mdss/8/) job passed.

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-container-add_mdss

rishabh-d-dave commented:
[non-container](https://2.jenkins.ceph.com/job/ceph-ansible-prs-dev-centos-non_container-add_mdss/13/) as well as [container](https://2.jenkins.ceph.com/job/ceph-ansible-prs-dev-centos-container-add_mdss/11/) version passed on CI.

rishabh-d-dave commented:
Rebasing and putting this PR to test on CI.

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-container-add_mdss

rishabh-d-dave commented:
During last to last push I added the commit that would use `site.yml.sample` and `site-container.yml.sample` instead of `add-mds.yml` in, both, containerized and non-containerized scenario and, both, [`dev-centos-container-add_mdss`](https://2.jenkins.ceph.com/job/ceph-ansible-prs-dev-centos-container-add_mdss/12/) and [`dev-centos-non_container-add_mdss`](https://2.jenkins.ceph.com/job/ceph-ansible-prs-dev-centos-non_container-add_mdss/14/) ran successfully on my machine as well as on GitHub CI.


During last push I moved around some code, removed add-mds.yml from the commit, made changes to tox.ini and squashed my commits.

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
jenkins test dev-centos-container-add_mdss

rishabh-d-dave commented:
@guits why did you add "[skip ci]" to the title?

guits commented:
@rishabh-d-dave because your branch wasn't (and still not) up to date and you are still pushing changes on this branch.
This triggers the CI for nothing since it's won't be merged until it's up to date with master.
We have a very low number of jenkins slaves, we can't afford to waste precious CI resources like this.

guits commented:
jenkins test dev-centos-container-add_mdss

guits commented:
jenkins test dev-centos-non_container-add_mdss

guits commented:
jenkins test dev-centos-non_container-switch_to_containers

guits commented:
jenkins test dev-centos-non_container-add_mdss

guits commented:
jenkins test dev-centos-container-add_mdss

guits commented:
jenkins test dev-centos-non_container-switch_to_containers

rishabh-d-dave commented:
jenkins test dev-centos-non_container-switch_to_containers

guits commented:
jenkins test dev-centos-container-add_mdss

guits commented:
jenkins test dev-centos-non_container-add_mdss

rishabh-d-dave commented:
Rebased and got containerized scenario working.

guits commented:
jenkins test dev-centos-non_container-all_daemons

guits commented:
jenkins test dev-centos-container-all_daemons

guits commented:
jenkins test dev-centos-container-all_daemons

guits commented:
jenkins test dev-centos-non_container-all_daemons

guits commented:
jenkins test dev-centos-non_container-add_mdss

guits commented:
jenkins test dev-centos-container-add_mdss
